src/raster/buffer.rs

```diff
@@ -0,0 +1,86 @@
+use crate::raster::GdalType;
+
+#[cfg(feature = "ndarray")]
+use ndarray::Array2;
+
+/// A 2-D array backed by it's `size` (cols, rows) and a row-major `Vec<T>` and it's dimensions.
+#[derive(Debug, Clone, PartialEq)]
+pub struct Buffer<T> {
+    pub size: (usize, usize),
+    pub data: Vec<T>,
+}
+
+impl<T: GdalType> Buffer<T> {
+    /// Construct a new buffer from `size` (`(cols, rows)`) and `Vec<T>`.
+    ///
+    /// # Notes
+    /// Assumes `size.0 * size.1 == data.len()`.
+    pub fn new(size: (usize, usize), data: Vec<T>) -> Buffer<T> {
+        debug_assert!(
+            size.0 * size.1 == data.len(),
+            "size {:?} does not match length {}",
+            size,
+            data.len()
+        );
+        Buffer { size, data }
+    }
+
+    #[cfg(feature = "ndarray")]
+    /// Convert `self` into an [`ndarray::Array2`].
+    pub fn to_array(self) -> crate::errors::Result<Array2<T>> {
+        // Array2 shape is (rows, cols) and Buffer shape is (cols in x-axis, rows in y-axis)
+        Ok(Array2::from_shape_vec(
+            (self.size.1, self.size.0),
+            self.data,
+        )?)
+    }
+}
+
+pub type ByteBuffer = Buffer<u8>;
+
+#[cfg(feature = "ndarray")]
+impl<T: GdalType> TryFrom<Buffer<T>> for Array2<T> {
+    type Error = crate::errors::GdalError;
+
+    fn try_from(value: Buffer<T>) -> Result<Self, Self::Error> {
+        value.to_array()
+    }
+}
+
+#[cfg(feature = "ndarray")]
+impl<T: GdalType + Copy> From<Array2<T>> for Buffer<T> {
+    fn from(value: Array2<T>) -> Self {
+        // Array2 shape is (rows, cols) and Buffer shape is (cols in x-axis, rows in y-axis)
+        let shape = value.shape().to_vec();
```

Comment:
No need to `to_vec()` here, `shape` appears to be a `&[usize]`.

Reply:
I got an ownership error without that.

Reply:
A `let (cols, rows) = (shape[1], shape[0]);` should fix that.